File: swift-bridge/Sources/AudioUnitBridge/AVAudioUnitMIDIInstrument.swift
import AudioToolbox
import AVFAudio
import CoreMIDI
import Foundation

func encodeAvAudioUnitMIDIInstrument(_ instrument: AVAudioUnitMIDIInstrument) -> [String: Any] {
    var dict = encodeAvAudioUnit(instrument)
    dict["supportsMIDIEventList"] = true
    return dict
}

@_cdecl("au_av_midi_instrument_create")
public func au_av_midi_instrument_create(
    _ type: UInt32,
    _ subtype: UInt32,
    _ manufacturer: UInt32,
    _ flags: UInt32,
    _ flagsMask: UInt32,
    _ outUnit: UnsafeMutablePointer<UnsafeMutableRawPointer?>,
    _ outErrorMsg: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>
) -> Int32 {
    outUnit.pointee = nil
    outErrorMsg.pointee = nil
    let desc = makeDesc(type, subtype, manufacturer, flags, flagsMask)
    let instrument = AVAudioUnitMIDIInstrument(audioComponentDescription: desc)
    outUnit.pointee = retainBox(instrument)
    return AU_OK
}

@_cdecl("au_av_midi_instrument_release")
public func au_av_midi_instrument_release(_ ptr: UnsafeMutableRawPointer) {
    releaseBox(ptr, as: AVAudioUnitMIDIInstrument.self)
}

@_cdecl("au_av_midi_instrument_snapshot_json")
public func au_av_midi_instrument_snapshot_json(_ ptr: UnsafeMutableRawPointer) -> UnsafeMutablePointer<CChar>? {
    jsonCString(encodeAvAudioUnitMIDIInstrument(borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self)))
}

@_cdecl("au_av_midi_instrument_as_avunit")
public func au_av_midi_instrument_as_avunit(_ ptr: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer {
    retainBox(borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self) as AVAudioUnit)
}

@_cdecl("au_av_midi_instrument_start_note")
public func au_av_midi_instrument_start_note(_ ptr: UnsafeMutableRawPointer, _ note: UInt8, _ velocity: UInt8, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).startNote(note, withVelocity: velocity, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_stop_note")
public func au_av_midi_instrument_stop_note(_ ptr: UnsafeMutableRawPointer, _ note: UInt8, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).stopNote(note, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_send_controller")
public func au_av_midi_instrument_send_controller(_ ptr: UnsafeMutableRawPointer, _ controller: UInt8, _ value: UInt8, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendController(controller, withValue: value, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_send_pitch_bend")
public func au_av_midi_instrument_send_pitch_bend(_ ptr: UnsafeMutableRawPointer, _ value: UInt16, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendPitchBend(value, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_send_pressure")
public func au_av_midi_instrument_send_pressure(_ ptr: UnsafeMutableRawPointer, _ pressure: UInt8, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendPressure(pressure, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_send_pressure_for_key")
public func au_av_midi_instrument_send_pressure_for_key(_ ptr: UnsafeMutableRawPointer, _ key: UInt8, _ value: UInt8, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendPressure(forKey: key, withValue: value, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_send_program_change")
public func au_av_midi_instrument_send_program_change(_ ptr: UnsafeMutableRawPointer, _ program: UInt8, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendProgramChange(program, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_send_program_change_bank")
public func au_av_midi_instrument_send_program_change_bank(_ ptr: UnsafeMutableRawPointer, _ program: UInt8, _ bankMSB: UInt8, _ bankLSB: UInt8, _ channel: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendProgramChange(program, bankMSB: bankMSB, bankLSB: bankLSB, onChannel: channel)
}

@_cdecl("au_av_midi_instrument_send_midi_event2")
public func au_av_midi_instrument_send_midi_event2(_ ptr: UnsafeMutableRawPointer, _ status: UInt8, _ data1: UInt8, _ data2: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendMIDIEvent(status, data1: data1, data2: data2)
}

@_cdecl("au_av_midi_instrument_send_midi_event1")
public func au_av_midi_instrument_send_midi_event1(_ ptr: UnsafeMutableRawPointer, _ status: UInt8, _ data1: UInt8) {
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendMIDIEvent(status, data1: data1)
}

@_cdecl("au_av_midi_instrument_send_sysex")
public func au_av_midi_instrument_send_sysex(_ ptr: UnsafeMutableRawPointer, _ bytes: UnsafePointer<UInt8>?, _ length: Int) {
    let data: Data
    if let bytes {
        data = Data(bytes: bytes, count: length)
    } else {
        data = Data()
    }
    borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).sendMIDISysExEvent(data)
}

@_cdecl("au_av_midi_instrument_send_event_list")
public func au_av_midi_instrument_send_event_list(_ ptr: UnsafeMutableRawPointer, _ list: UnsafeRawPointer?) {
    guard let list else { return }
    list.assumingMemoryBound(to: MIDIEventList.self).withMemoryRebound(to: MIDIEventList.self, capacity: 1) { rebound in
        borrowBox(ptr, as: AVAudioUnitMIDIInstrument.self).send(rebound)
    }
}

File: src/av_audio_unit_midi_instrument.rs
//! `AVAudioUnitMIDIInstrument` wrappers.

use core::ffi::c_void;

use serde::Deserialize;

use crate::av_audio_unit::{AvAudioUnit, AvAudioUnitInfo};
use crate::component_description::AudioComponentDescription;
use crate::error::AuError;
use crate::ffi;
use crate::util::{status_result, take_json};

/// Snapshot of an `AVAudioUnitMIDIInstrument`.
#[derive(Debug, Clone, Deserialize)]
pub struct AvAudioUnitMidiInstrumentInfo {
    #[serde(flatten)]
    pub av_audio_unit: AvAudioUnitInfo,
    #[serde(rename = "supportsMIDIEventList")]
    pub supports_midi_event_list: bool,
}

/// Owned handle to an `AVAudioUnitMIDIInstrument`.
pub struct AvAudioUnitMidiInstrument {
    ptr: *mut c_void,
}

/// Backwards-friendly alias preserving Apple's acronym casing.
pub type AvAudioUnitMIDIInstrument = AvAudioUnitMidiInstrument;

unsafe impl Send for AvAudioUnitMidiInstrument {}

impl Drop for AvAudioUnitMidiInstrument {
    fn drop(&mut self) {
        unsafe { ffi::au_av_midi_instrument_release(self.ptr) };
    }
}

impl AvAudioUnitMidiInstrument {
    /// Create a MIDI instrument for the given component description.
    pub fn new(description: AudioComponentDescription) -> Result<Self, AuError> {
        let mut unit_ptr = core::ptr::null_mut();
        let mut error_ptr = core::ptr::null_mut();
        let status = unsafe {
            ffi::au_av_midi_instrument_create(
                description.component_type,
                description.component_subtype,
                description.component_manufacturer,
                description.component_flags,
                description.component_flags_mask,
                &mut unit_ptr,
                &mut error_ptr,
            )
        };
        status_result(status, error_ptr)?;
        Ok(Self { ptr: unit_ptr })
    }

    /// Snapshot the instrument metadata.
    pub fn info(&self) -> Result<AvAudioUnitMidiInstrumentInfo, AuError> {
        let ptr = unsafe { ffi::au_av_midi_instrument_snapshot_json(self.ptr) };
        take_json(ptr)
    }

    /// Clone the base `AVAudioUnit` handle.
    pub fn as_av_audio_unit(&self) -> AvAudioUnit {
        let ptr = unsafe { ffi::au_av_midi_instrument_as_avunit(self.ptr) };
        AvAudioUnit::from_raw(ptr)
    }

    /// Send a MIDI note-on event.
    pub fn start_note(&self, note: u8, velocity: u8, channel: u8) {
        unsafe { ffi::au_av_midi_instrument_start_note(self.ptr, note, velocity, channel) };
    }

    /// Send a MIDI note-off event.
    pub fn stop_note(&self, note: u8, channel: u8) {
        unsafe { ffi::au_av_midi_instrument_stop_note(self.ptr, note, channel) };
    }

    /// Send a MIDI controller event.
    pub fn send_controller(&self, controller: u8, value: u8, channel: u8) {
        unsafe { ffi::au_av_midi_instrument_send_controller(self.ptr, controller, value, channel) };
    }

    /// Send a pitch-bend event.
    pub fn send_pitch_bend(&self, value: u16, channel: u8) {
        unsafe { ffi::au_av_midi_instrument_send_pitch_bend(self.ptr, value, channel) };
    }

    /// Send channel pressure.
    pub fn send_pressure(&self, pressure: u8, channel: u8) {
        unsafe { ffi::au_av_midi_instrument_send_pressure(self.ptr, pressure, channel) };
    }

    /// Send polyphonic key pressure.
    pub fn send_pressure_for_key(&self, key: u8, value: u8, channel: u8) {
        unsafe {
            ffi::au_av_midi_instrument_send_pressure_for_key(
                self.ptr, key, value, channel,
            );
        };
    }

    /// Send a program-change event.
    pub fn send_program_change(&self, program: u8, channel: u8) {
        unsafe { ffi::au_av_midi_instrument_send_program_change(self.ptr, program, channel) };
    }

    /// Send a program-change plus bank-select event.
    #[allow(clippy::similar_names)]
    pub fn send_program_change_bank(
        &self,
        program: u8,
        bank_msb_value: u8,
        bank_lsb_value: u8,
        channel: u8,
    ) {
        unsafe {
            ffi::au_av_midi_instrument_send_program_change_bank(
                self.ptr,
                program,
                bank_msb_value,
                bank_lsb_value,
                channel,
            );
        };
    }

    /// Send a two-byte MIDI event.
    pub fn send_midi_event2(&self, status: u8, data1: u8, data2: u8) {
        unsafe { ffi::au_av_midi_instrument_send_midi_event2(self.ptr, status, data1, data2) };
    }

    /// Send a one-byte MIDI event.
    pub fn send_midi_event1(&self, status: u8, data1: u8) {
        unsafe { ffi::au_av_midi_instrument_send_midi_event1(self.ptr, status, data1) };
    }

    /// Send a `SysEx` event.
    pub fn send_sysex(&self, bytes: &[u8]) {
        unsafe { ffi::au_av_midi_instrument_send_sysex(self.ptr, bytes.as_ptr(), bytes.len()) };
    }

    /// Forward a raw `MIDIEventList` pointer to the instrument.
    ///
    /// # Safety
    /// The caller must ensure the pointer refers to a valid `MIDIEventList`.
    pub unsafe fn send_event_list_raw(&self, list: *const c_void) {
        ffi::au_av_midi_instrument_send_event_list(self.ptr, list);
    }
}
